File: packages-gen/js/event/util-throttle.js
const xtd = require("@xtalk/lang/common-data.js")

function throttle_create(handler,now_fn){
  return {
    "now_fn":(null == now_fn) ? (function (){
        return Date.now();
      }) : now_fn,
    "handler":handler,
    "active":{},
    "queued":{}
  };
}

function throttle_run_async(throttle,id,args){
  let {active,handler,queued} = throttle;
  let key = String(id);
  args = xtd.arrayify(args);
  let inputs = [id];
  for(let arg of args){
    inputs.push(arg);
  };
  let base_promise = Promise.resolve().then(function (){
    return handler.apply(null,inputs);
  });
  return base_promise.finally(function (){
    delete(active[key]);
    let qentry = queued[key];
    if(null != qentry){
      active[key] = qentry;
      delete(queued[key]);
      throttle_run_async(throttle,id,qentry["args"]);
    }
  });
}

function throttle_run(throttle,id,args){
  let {active,now_fn,queued} = throttle;
  let key = String(id);
  args = xtd.arrayify(args);
  let qentry = queued[key];
  if(null != qentry){
    return qentry;
  }
  let aentry = active[key];
  if(null != aentry){
    qentry = {
      "promise":aentry["promise"],
      "started":now_fn(),
      "args":aentry["args"]
    };
    queued[key] = qentry;
    return qentry;
  }
  let promise = throttle_run_async(throttle,id,args);
  aentry = {"promise":promise,"started":now_fn(),"args":args};
  active[key] = aentry;
  return aentry;
}

function throttle_waiting(throttle){
  let {active,queued} = throttle;
  return xtd.arr_union(Object.keys(active),Object.keys(queued));
}

function throttle_active(throttle){
  let {active} = throttle;
  return Object.keys(active);
}

function throttle_queued(throttle){
  let {queued} = throttle;
  return Object.keys(queued);
}

module.exports = {
  ["throttle_create"]:throttle_create,
  ["throttle_run_async"]:throttle_run_async,
  ["throttle_run"]:throttle_run,
  ["throttle_waiting"]:throttle_waiting,
  ["throttle_active"]:throttle_active,
  ["throttle_queued"]:throttle_queued
}

File: packages-gen/dart/libs/xt-event/lib/util-throttle.dart
import 'package:xtalk_lang/common-data.dart' as xtd;
import 'dart:async';


throttle_create(handler, now_fn) {
  return <dynamic, dynamic>{
    "now_fn":(null == now_fn) ? (() {
        return DateTime.now().millisecondsSinceEpoch;
      }) : now_fn,
    "handler":handler,
    "active":<dynamic, dynamic>{},
    "queued":<dynamic, dynamic>{}
  };
}

throttle_run_async(throttle, id, args) {
  var active = throttle["active"];
  var handler = throttle["handler"];
  var queued = throttle["queued"];
  var key = (id).toString();
  args = xtd.arrayify(args);
  var inputs = <dynamic>[id];
  var arr_50713 = args;
  for(var i50714 = 0; i50714 < arr_50713.length; ++i50714){
    var arg = arr_50713[i50714];
    inputs.add(arg);
  };
  var base_promise = Future.sync(() {
    return Function.apply(handler,inputs);
  });
  return ((Future.sync(() => base_promise)) as Future<dynamic>).whenComplete(() async { await Function.apply(() {
    active.remove(key);
    var qentry = queued[key];
    if(null != qentry){
      active[key] = qentry;
      queued.remove(key);
      throttle_run_async(throttle,id,qentry["args"]);
    }
  },<dynamic>[]); });
}

throttle_run(throttle, id, args) {
  var active = throttle["active"];
  var now_fn = throttle["now_fn"];
  var queued = throttle["queued"];
  var key = (id).toString();
  args = xtd.arrayify(args);
  var qentry = queued[key];
  if(null != qentry){
    return qentry;
  }
  var aentry = active[key];
  if(null != aentry){
    qentry = <dynamic, dynamic>{
      "promise":aentry["promise"],
      "started":Function.apply((now_fn as Function),<dynamic>[]),
      "args":aentry["args"]
    };
    queued[key] = qentry;
    return qentry;
  }
  aentry = <dynamic, dynamic>{
    "promise":null,
    "started":Function.apply((now_fn as Function),<dynamic>[]),
    "args":args
  };
  active[key] = aentry;
  var promise = throttle_run_async(throttle,id,args);
  aentry["promise"] = promise;
  return aentry;
}

throttle_waiting(throttle) {
  var active = throttle["active"];
  var queued = throttle["queued"];
  return xtd.arr_union(
    List<dynamic>.from(( active ).keys),
    List<dynamic>.from(( queued ).keys)
  );
}

throttle_active(throttle) {
  var active = throttle["active"];
  return List<dynamic>.from(( active ).keys);
}

throttle_queued(throttle) {
  var queued = throttle["queued"];
  return List<dynamic>.from(( queued ).keys);
}